cfi/lib/src/cfi_ctr.rs

```diff
@@ -0,0 +1,204 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+    cfi_ctr.rs
+
+Abstract:
+
+    File contains CFI Integer and Counter implementations. The counter is based on ideas from
+    Trusted Firmware-M firmware.
+
+References:
+    https://tf-m-user-guide.trustedfirmware.org/design_docs/tfm_physical_attack_mitigation.html
+
+--*/
+
+use crate::cfi::{cfi_panic, CfiPanicInfo};
+use crate::xoshiro::{Xoshiro128, Xoshiro128Reg};
+#[cfg(not(feature = "cfi-test"))]
+use caliptra_common::memory_layout::{CFI_MASK_ORG, CFI_VAL_ORG};
+use core::default::Default;
+
+#[cfg(feature = "cfi-test")]
+static mut CFI_VAL: u32 = 0u32;
+#[cfg(feature = "cfi-test")]
+static mut CFI_MASK: u32 = 0u32;
+
+/// CFI Integer
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub struct CfiInt {
+    /// Actual Value
+    val: u32,
+
+    /// Masked Value
+    masked_val: u32,
+}
+
+impl CfiInt {
+    /// Integer mask with high hamming distance
+    const MASK: u32 = 0xA5A5A5A5;
+
+    /// Create integer from raw values
+    fn from_raw(val: u32, masked_val: u32) -> Self {
+        Self { val, masked_val }
+    }
+
+    /// Encode the integer
+    fn encode(val: u32) -> Self {
+        Self {
+            val,
+            masked_val: val ^ Self::MASK,
+        }
+    }
+
+    /// Check if the integer is valid
+    fn is_valid(&self) -> bool {
+        self.val == self.masked_val ^ Self::MASK
+    }
+}
+
+impl Default for CfiInt {
+    /// Returns the "default value" for a type.
+    fn default() -> Self {
+        Self::encode(0)
+    }
+}
+
+/// CFI counter
+pub enum CfiCounter {}
+
+impl CfiCounter {
+    /// Reset counter
+    #[inline(never)]
+    pub fn reset() {
+        Xoshiro128::new(Xoshiro128Reg);
+        Self::write(CfiInt::default());
+    }
+
+    // Zero the counter
+    pub fn corrupt() {
+        Self::write(CfiInt {
+            val: 0,
+            masked_val: 0,
+        });
+    }
+
+    /// Increment counter
+    #[inline(never)]
+    pub fn increment() -> CfiInt {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            let int = Self::read();
+            if !int.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+
+            let (new, overflow) = int.val.overflowing_add(1);
+            if overflow {
+                cfi_panic(CfiPanicInfo::CounterOverflow);
```

Comment:
Where do we ensure that we never init the counter to a value where the overflow will happen naturally?

Reply:
It's done at the start of rom_entry function in rom/dev/src/main.rs. The counter value that we start with is 0xA5A5A5A5 (0 XORed 0xA5A5A5A5).